Add upsert ttl configs for Pinot upsert optimizations

deemoliu opened the pull request:
Part of a series of PRs for Pinot Upsert Optimizations.

Issue: #9529 
Tags: `feature` `release-notes`

Made the following change to upsertConfig
- Add upsert ttl configs for Pinot upsert optimizations

Made the following changes to PartitionUpsertMetadataManager
- Clean up keys in primary key indexes during segment replacement (committing)

The change is proposed by [Upsert TTL feature design doc](https://docs.google.com/document/d/1AhkZPkf4St3hj96IXyNs6NOiItxm3LXw3NAgNbG3oRo/edit?pli=1#heading=h.z1lc2k9b0z02)

navina commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertTTLConfig.java:
```diff
@@ -0,0 +1,54 @@
+import javax.annotation.Nullable;
+
+
+public class UpsertTTLConfig {
+
+  @JsonPropertyDescription("ttl time unit")
+  private String _ttlTimeUnit;
+  @JsonPropertyDescription("ttl time value")
+  private String _ttlTimeValue;
```
1. can you include java docs for this class
2. It will be useful to include the supported time units in the `@JsonPropertyDescription` for `_ttlTimeUnit`

deemoliu replied:
@navina This PR is still in WIP status. Let me add the java docs and test coverage. 

thanks for code review!

navina commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/TTLPartitionUpsertMetadataManager.java:
```diff
@@ -0,0 +1,68 @@
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+@ThreadSafe
+public class TTLPartitionUpsertMetadataManager extends ConcurrentMapPartitionUpsertMetadataManager {
```
why is this class needed when you can just override the method `removeExpiredPrimaryKeys` in `ConcurrentMapPartitionUpsertMetadataManager`? 
Same question on `TTLTableUpsertMetadataManager`.

deemoliu replied:
@navina I will add more methods later in the TTLPartitionUpsertMetadataManager.

As mentioned in https://github.com/apache/pinot/issues/9529,

- When committing segment, update replaceSegment to clean up keys;
- when loading segments, get snapshot from peer-server to avoid re-compute

navina replied:
You can add more methods to it. But my point was that this TTL feature will only work with `ConcurrentMapPartitionUpsertMetadataManager` and none of the other implementations of `PartitionUpsertMetadataManager`. I don't think that was the intention in the design. Please correct me if I am missing something here. 

It will be better to model this as a separate component or a wrapper around any of the `PartitionUpsertMetadataManager`.

Jackie-Jiang replied:
+1 here. I think TTL should be a feature supported by general upsert metadata manager. Making it a separate class will prevent other storage format from supporting TTL

deemoliu replied:
thanks @navina and @Jackie-Jiang Let me refactor this part.

navina commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/UpsertTTLStrategy.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.concurrent.TimeUnit;
+
+
+public class UpsertTTLStrategy {
```
can you add javadocs for this class? 

IIuc, this seems to be a class to hold the values from `UpsertTTLConfig` with an additional method to convert to millis. So, is this class really needed?

deemoliu replied:
@navina this class is similar to TimeRetentionStrategy in package org.apache.pinot.controller.helix.core.retention.strategy. I added a new class so it can be extended in the future.
E.g. skipLate event etc. Different approach for expired primary keys cleanups, etc.

Jackie-Jiang replied:
I agree with @navina that this extra class is not really needed, unless we want different TTL handling strategies, at which time an interface should be introduced, instead of a concrete class.

deemoliu replied:
@navina @Jackie-Jiang gotcha. Let me merge adding an addition method in upsertTTLConfig and remove this class for now. I can introduce addition interface if needed in future. e.g. two map rotated approach for TTL handling.

navina commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:
```diff
@@ -162,6 +165,9 @@ protected abstract void addOrReplaceSegment(ImmutableSegmentImpl segment, Thread
       Iterator<RecordInfo> recordInfoIterator, @Nullable IndexSegment oldSegment,
       @Nullable MutableRoaringBitmap validDocIdsForOldSegment);
 
+  protected abstract void removeExpiredPrimaryKeys(IndexSegment segment, Iterator<RecordInfo> recordInfoIterator,
```
Can you add some javadocs for this new method? 

Also, I think this method should be included as a part of the `PartitionUpsertMetadataManager` interface itself since it manipulates the partition metadata. Is there a reason why it is not included as a part of the interface?

deemoliu replied:
@navina this method is similar to `addOrReplaceSegment()` method that is being called internally by replaceSegment.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:
```diff
@@ -217,6 +226,17 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
+        // remove expired primary keys from primary key indexes.
+        if (_upsertTTLStrategy != null) {
+          removeExpiredPrimaryKeys(segment, recordInfoIterator, _upsertTTLStrategy.getTtlInMs());
+          MutableRoaringBitmap validDocIdsSnapshot =
```
I don't follow the logic here. Why do we persist the valid doc ids here? The new added segment is almost always having all records valid.

deemoliu replied:
Thanks @Jackie-Jiang I think i can fix this logic by the following two changes.
- iterating primary key indexes to remove out-of-dated PKs, instead of iterating the recordInfoIterator 
- make sure the expired PK removal should be done before the new consuming segment might already be created.
Is there anything I missed.

deemoliu replied:
> Do we plan to implement the "Remove synchronously" approach described in the design doc? If so, it cannot be done in the replace segment because it is too late, and the new consuming segment might already be created. We will need to add a separate method which is invoked after the segment is getting committed, and before the new segment starts consuming.

@Jackie-Jiang is there some examples that invoked after the segment is getting committed, and before the new segment starts consuming?

Jackie-Jiang reviewed (commented):
Do we plan to implement the "Remove synchronously" approach described in the design doc? If so, it cannot be done in the replace segment because it is too late, and the new consuming segment might already be created. We will need to add a separate method which is invoked after the segment is getting committed, and before the new segment starts consuming.

chenboat commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -181,6 +184,32 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
+  /**
+   * When TTL is enabled for upsert, this function is used to remove expired keys from the primary key indexes.
+   *
+   * When committing consuming segment, we replaced the consuming segment with an immutable segments.
```
replaced-->replace

chenboat commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -181,6 +184,32 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
+  /**
+   * When TTL is enabled for upsert, this function is used to remove expired keys from the primary key indexes.
+   *
+   * When committing consuming segment, we replaced the consuming segment with an immutable segments.
```
Why do we want to do the cleanup of out-of-ttl keys during segment commit phase? This means the clean up is blocking and potentially slow down the ingestion. An alternative is to do it lazily as a background scheduled job. If I understand correctly, TTL for upsert table is an optimization, we do not need it synchronously. Here is the design doc about the choices

https://docs.google.com/document/d/1AhkZPkf4St3hj96IXyNs6NOiItxm3LXw3NAgNbG3oRo/edit?pli=1#heading=h.5ol3ncuq1lx2

Jackie-Jiang replied:
IIRC, we want to ensure the state for each replica to be the same. If we don't sync on the cleanup, each replica will potentially have different result, which can cause problem for upsert because during segment commit the segment can come from another server

deemoliu commented:
> Do we plan to implement the "Remove synchronously" approach described in the design doc? If so, it cannot be done in the replace segment because it is too late, and the new consuming segment might already be created. We will need to add a separate method which is invoked after the segment is getting committed, and before the new segment starts consuming.

Hi @Jackie-Jiang I checked this part, the function will be called during LLRealtimeSegmentDataManager.commitSegment(String, boolean). Since it's triggered when commit Segment, the new segment on the same partition should not started consuming. Can you help me understand why the new consuming segment might already be created?

Do you foresee any risk on the following operations during segment commit?

removeExpiredPrimaryKeys(segment, expiredTimestamp);
persistSnapshotForStableSegment(expiredTimestamp);

Jackie-Jiang commented:
> Hi @Jackie-Jiang I checked this part, the function will be called during LLRealtimeSegmentDataManager.commitSegment(String, boolean). Since it's triggered when commit Segment, the new segment on the same partition should not started consuming. Can you help me understand why the new consuming segment might already be created?
> 
> Do you foresee any risk on the following operations during segment commit?
> 
> removeExpiredPrimaryKeys(segment, expiredTimestamp); persistSnapshotForStableSegment(expiredTimestamp);

Only the committing server will call `commitSegment()`, but we need all servers to clean up the expired keys after finishing the current consuming segment and before starting the new consuming segment.

deemoliu commented:
> > Hi @Jackie-Jiang I checked this part, the function will be called during LLRealtimeSegmentDataManager.commitSegment(String, boolean). Since it's triggered when commit Segment, the new segment on the same partition should not started consuming. Can you help me understand why the new consuming segment might already be created?
> > Do you foresee any risk on the following operations during segment commit?
> > removeExpiredPrimaryKeys(segment, expiredTimestamp); persistSnapshotForStableSegment(expiredTimestamp);
> 
> Only the committing server will call `commitSegment()`, but we need all servers to clean up the expired keys after finishing the current consuming segment and before starting the new consuming segment.

Hi @Jackie-Jiang, thanks for the commit. Can you help understand this part? 
In my understanding, each partitionManager has its own PrimaryKeyToRecordLocation map. When committing a new segment in one partition, do we need other partitionManagers to clean up the expired keys?

Jackie-Jiang commented:
> Hi @Jackie-Jiang, thanks for the commit. Can you help understand this part? In my understanding, each partitionManager has its own PrimaryKeyToRecordLocation map. When committing a new segment in one partition, do we need other partitionManagers to clean up the expired keys?

No, each partition can be handled independently.

deemoliu commented:
> > Hi @Jackie-Jiang, thanks for the commit. Can you help understand this part? In my understanding, each partitionManager has its own PrimaryKeyToRecordLocation map. When committing a new segment in one partition, do we need other partitionManagers to clean up the expired keys?
> 
> No, each partition can be handled independently.

Discussed with Jackie, and found the following issue.

There are multiple servers consuming the same segment, only one of them will commit the segment, and the other servers will download the committed segment. With the current approach, the servers not committing the segment won't clean up the expired keys, which will cause inconsistency.

Basically we want to make sure all servers clean up the expired keys after finishing the current consuming segment but before starting the next consuming segment





[12:17](https://apache-pinot.slack.com/archives/D01CCL54XCM/p1677183435873319)
We want to block the consumption until the keys are cleaned up

deemoliu commented:
> > Hi @Jackie-Jiang I checked this part, the function will be called during LLRealtimeSegmentDataManager.commitSegment(String, boolean). Since it's triggered when commit Segment, the new segment on the same partition should not started consuming. Can you help me understand why the new consuming segment might already be created?
> > Do you foresee any risk on the following operations during segment commit?
> > removeExpiredPrimaryKeys(segment, expiredTimestamp); persistSnapshotForStableSegment(expiredTimestamp);
> 
> Only the committing server will call `commitSegment()`, but we need all servers to clean up the expired keys after finishing the current consuming segment and before starting the new consuming segment.

thanks @Jackie-Jiang @chenboat 
I addressed the comments, please take a look.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:
```diff
@@ -395,6 +399,31 @@ protected void finishOperation() {
+   */
+  @Override
+  public void removeExpiredPrimaryKeys(Comparable timestamp) {
+    if (_upsertTTLConfig != null && _upsertTTLConfig.getTtlInMs() > 0) {
```
This check is redundant because it can only be invoked when TTL is configured. We may remove the default implementation. Same for `persistSnapshotForStableSegment()`

chenboat replied:
@deemoliu can you address this comment?

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:
```diff
@@ -57,12 +57,18 @@ public enum Strategy {
   private boolean _enableSnapshot;
 
+  @JsonPropertyDescription("Whether to use TTL to reduce upsert metadata memory footprint")
+  private boolean _enableTTL;
```
Suggest removing this. Providing `_upsertTTLConfig` implicitly indicate that TTL is enabled

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertTTLConfig.java:
```diff
@@ -0,0 +1,78 @@
+public class UpsertTTLConfig {
+
+  @JsonPropertyDescription("ttl time unit, supported time units are DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS")
+  private String _ttlTimeUnit;
```
We should be able to directly declare it as `TimeUnit` type

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertTTLConfig.java:
```diff
@@ -0,0 +1,78 @@
+  @JsonPropertyDescription("ttl time unit, supported time units are DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS")
+  private String _ttlTimeUnit;
+  @JsonPropertyDescription("ttl time value")
+  private String _ttlTimeValue;
```
Declare it as `long` type

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -267,4 +333,28 @@ public Comparable getComparisonValue() {
+  @VisibleForTesting
+  static class SegmentInfo implements Comparable<SegmentInfo> {
+    private final IndexSegment _segment;
+    private final long _endtime;
```
```suggestion
    private final long _endTimeMs;
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -53,14 +56,19 @@ public class ConcurrentMapPartitionUpsertMetadataManager extends BasePartitionUp
 
+  // keep track of all segments that haven't reach stable state to persist snapshot
+  @VisibleForTesting
+  final PriorityBlockingQueue<SegmentInfo> _nonPersistedSegmentsQueue = new PriorityBlockingQueue<>();
```
We can construct it only when TTL is enabled, and in the following places just check if it is `null`

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -181,6 +198,55 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
+    _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
+      assert recordLocation.getComparisonValue() != null;
+      if (recordLocation.getComparisonValue().compareTo(expiredTimestamp) < 0) {
+        _primaryKeyToRecordLocationMap.remove(primaryKey);
```
This can cause race condition. You want to do check and remove in case the record is updated during remove
```suggestion
        _primaryKeyToRecordLocationMap.remove(primaryKey, recordLocation);
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -181,6 +198,55 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
+   * @return void
+   */
+  @Override
+  public void doRemoveExpiredPrimaryKeys(Comparable expiredTimestamp) {
```
Currently this works only when the comparison column is timestamp (time column in milliseconds). We should add some validation to the table config

KKcorps replied:
Yes, plus now we have support for multiple comparison columns as well. We can enforce a requirement for now that if TTL is enabled only one comparison column can be there and it needs to be timestamp.

navina replied:
+1 to the above. can you please call this out in the javadoc and in the oss pinot docs (when you get to it)?

deemoliu replied:
Sure, will do

Jackie-Jiang reviewed (commented):
We need to add more controls to the snapshots. We should mark snapshot persisted by TTL differently from the regular snapshot:
- When adding segment, we don't want to add metadata for segment with TTL snapshot
- We don't want to persist regular snapshot when TTL snapshot exists
- Do we want to allow TTL without enabling snapshot
- What if user want to change the TTL value or disable TTL

deemoliu commented:
> We need to add more controls to the snapshots. We should mark snapshot persisted by TTL differently from the regular snapshot:
> 
> * When adding segment, we don't want to add metadata for segment with TTL snapshot
> * We don't want to persist regular snapshot when TTL snapshot exists
> * Do we want to allow TTL without enabling snapshot
> * What if user want to change the TTL value or disable TTL

Thanks @Jackie-Jiang I tried to solve the above comment with the following approach, can you please review?
(1) When adding segment, we don't want to add metadata for segment with TTL snapshot
The current implementation doesn't add metadata for segment with TTL snapshot.

(2) We don't want to persist regular snapshot when TTL snapshot exists
```
  public static final String VALID_DOC_IDS_SNAPSHOT_FILE_NAME = "validdocids.bitmap.snapshot";
  public static final String VALID_DOC_IDS_TTL_SNAPSHOT_FILE_NAME = "validdocids.bitmap.ttl.snapshot";
```
I used different snapshot name for "snapshotOnly" and "TTLEnabled", thus it's okay to persist regular snapshot when TTL snapshot exists. Do you think we still need to add this constraint in the tableConfig?

(3)  Do we want to allow TTL without enabling snapshot
We can allow enable or disable `snapshotOnly` with TTL feature, since they use different snapshots.
Currently if TTL is enabled, the `TTLEnabled` snapshot for old segments will automatically generated during  new segment sealing. 

(4) Change the TTL value and disable TTL can be supported by re-initalized ConcurrentHashmapPartitionUpsertManager. However there need to be another PR to handle some cases.
e.g. 
```
    while (_nonPersistedSegmentsQueue.peek()._endTimeMs < expiredTimestamp) {
      IndexSegment segment = _nonPersistedSegmentsQueue.poll()._segment;
      ...
```
TTL value change can lead to changing to expiredTimestamp, and so we might see missing validDocsSnapshot for some segments. For now. I will add a validation to avoid updating the TTL values (user need to create new table if they need to change TTL value).

KKcorps commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java:
```diff
@@ -159,7 +159,11 @@ public void deleteValidDocIdsSnapshot() {
-  private File getValidDocIdsSnapshotFile() {
+  private File getValidDocIdsSnapshotFile(boolean isTTLEnabled) {
+    if (isTTLEnabled) {
+      return new File(SegmentDirectoryPaths.findSegmentDirectory(_segmentMetadata.getIndexDir()),
```
what is the need for a seperate file? Both are doing the same thing right from the load perspective?

deemoliu replied:
thanks @KKcorps The reason to have a separate file is that We should mark snapshot persisted by TTL differently from the regular snapshot, so validDocsIdsSnapshot from upsertTTL will not be override by SnapshotEnabled feature. We can also use snapshot for segment compaction and other upsert improvements.

Another potential solution is disable `enableSnapshot` for upsert TTL so that both snapshots won't overwrite each other.

KKcorps replied:
Makes sense. Then I want to understand, how does this process work during server restart because we will be loading validDocIds from the snapshot file and then discarding them based on this ttl_snapshot file?

Also, what happens if all docs of the segment have been removed due to TTL and we restart the server?

deemoliu replied:
@KKcorps thanks for the code review and pointing out this part. 
I think I missed this part in the code, `loadValidDocIdsFromSnapshot(true)` and `deleteValidDocIdsSnapshot(true)` should be called here when TTLConfig exists. I will add it.

I also think about this part again, I feel snapshot feature and TTLConfig  and enableSnapshot might not need to be enabled together, otherwise we will store duplicated snapshots, and the snapshot persisted by the snapshotEnabled will never be used by TTL feature. If we go through this route, we will handle three scenarios 
1) TTL config enabled
2) snapshot enabled
3) both are not enabled.

I will also add validation to avoid both config enabled in upsertConfig.

deemoliu replied:
With this change, the load and deletion behavior will be the same for snapshotEnabled and upsertTTL upon server restart, however how to persist snaphost is different for the above two features.

KKcorps commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:
```diff
@@ -73,13 +75,15 @@ public abstract class BasePartitionUpsertMetadataManager implements PartitionUps
   protected BasePartitionUpsertMetadataManager(String tableNameWithType, int partitionId,
       List<String> primaryKeyColumns, List<String> comparisonColumns, HashFunction hashFunction,
-      @Nullable PartialUpsertHandler partialUpsertHandler, boolean enableSnapshot, ServerMetrics serverMetrics) {
+      @Nullable PartialUpsertHandler partialUpsertHandler, @Nullable UpsertTTLConfig upsertTTLConfig,
```
We are cramming too many args in this constructor. Should directly pass upsert config here. If out of scope for this PR, please create another issue for this.

deemoliu replied:
@KKcorps sure, will do in another PR.

chenboat commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:
```diff
@@ -181,6 +203,60 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
+   *
+   * When committing consuming segment, we replace the consuming segment with an immutable segments.
+   * After replaceSegment, we iterate over recordInfoIterator to find validDocIds that are expired (out-of-TTL).
+   * Primarykey expired when the comparison time value of the record is less or equal to (segmentEndTime - TTL).
```
what is segmentEndTime? why use it instead of the current system time?

deemoliu replied:
@chenboat we have to use eventTime instead of system for primary key cleanup / snapshot persistant.

chenboat commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:
```diff
@@ -526,6 +529,32 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
+  /**
+   * Handle Upsert TTL during segment state change from CONSUMING to ONLINE.
+   *
+   * When a new segments commit, we will remove primary key indexes from heap that has timestamp < (eventTime - TTL).
```
change to "when a new segment commits"

chenboat replied:
Heap is too general. Change the term heap to the exact data structure used

deemoliu replied:
thanks, updated.

chenboat commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:
```diff
@@ -526,6 +529,32 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
+   * Handle Upsert TTL during segment state change from CONSUMING to ONLINE.
+   *
+   * When a new segments commit, we will remove primary key indexes from heap that has timestamp < (eventTime - TTL).
+   * For all segments sealed before (eventTime-TTL), their validDocIds won't get updated. We do an one-time persistance.
```
Please explain what you mean by "one-time persistance".

deemoliu replied:
Only segments that haven't been persisted before will do snapshot persistence.

We have a PQ to keep track all the segments that snapshot not have been persisted.

chenboat commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:
```diff
@@ -526,6 +529,32 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
+    long expiredTimestamp = endTimeInMillis - upsertTTLConfig.getTtlInMs();
+
+    partitionUpsertMetadataManager.removeExpiredPrimaryKeys(expiredTimestamp);
+    partitionUpsertMetadataManager.persistSnapshotForStableSegment(expiredTimestamp);
```
Should we use persistSnapshotForStableSegments because usually there are more than 1 stable segments per partition?

deemoliu replied:
gotcha, updated.

navina commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java:
```diff
@@ -84,6 +84,16 @@ public interface PartitionUpsertMetadataManager extends Closeable {
+  /**
+   * Remove from the primary key index when the PK are expired if TTL is enabled.
+   */
+  void removeExpiredPrimaryKeys(Comparable expiredTimestamp);
+
+  /**
+   * Persist validDocIds snapshot when the validDocIds is stable if TTL is enabled.
+   */
+  void persistSnapshotForStableSegments(long expiredTimestamp);
+
```
Why is the `expiredTimestamp` in `removeExpiredPrimaryKeys` a `Comparable` , whereas it is a `long` in `persistSnapshotForStableSegments` ?

deemoliu replied:
@navina The comparison column value is a comparable, in the primaryKey to recordLocation map.

deemoliu commented:
The PR after cleanup and review was merged https://github.com/apache/pinot/pull/10915
thanks everyone for design discussion and review!
